src/rc-socket.js

```diff
@@ -30,293 +21,341 @@ var root = this;
-  // Instance vars
-  this.protocols = protocols;
-  this.URL = url;
-
-  // Hack to fix firefox triggering close on leave
+  // Hack P1: Safegaurd against firefox behavior where close event is
+  // triggered on page navigation and results in an attempted reconnect.
   window.onbeforeunload = function () {
-    this.unload = true;
+    this.hasUnloaded = true;
   }.bind(this);
 
-  // Connect dAwG! - delay to add handlers
-  setTimeout(this.connect.bind(this), 0);
+  // Delay connect so that we can immediately add socket handlers.
+  setTimeout(this._connect.bind(this), 0);
 };
 
-
 /**
- * Wrapper around websocket.
+ * @publc
+ * @memberof RcSocket
  *
- * @private
- * @param {Number} reconnectAttempt - Name of event to fire.
- */
-RcSocket.prototype.connect = function () {
-  // New connecting websocket
-  if (this.protocols) {
-    this.ws = new WebSocket(this.URL, this.protocols);
-  } else {
-    this.ws = new WebSocket(this.URL);
-  }
-  
-  // Attach an id to the internal web socket. Could be use for various reasons
-  // but initially being introduced for debugging purposes.
-  this.ws.id = Date.now();
-
-  // Update
-  this._stateChanged('CONNECTING', 'onconnecting');
-
-  // Start timer
-  var hasConnected = false,
-      timeout      = this._setTimeout(this.ws);
-
-  /* ---------------------------------
-   * open
-   * -------------------------------*/
-  this.ws.onopen = function (evt) {
-    clearTimeout(timeout);
-
-    // Fix error where close is explicitly called
-    // but onopen event is still triggered
-    if (this.forced) {
-      return this.close();
-    }
-    
-    hasConnected = true;
-    this.attempts = 1;
-    this._stateChanged('OPEN', 'onopen', evt);
-    this._sendQueued();
-  }.bind(this);
-    
-  /* ---------------------------------
-   * close
-   * -------------------------------*/
-  this.ws.onclose = function (evt) {
-    clearTimeout(timeout);
-
-    this.ws = null;
-    if (this.forced) {
-      evt.forced = this.forced;
-      this._stateChanged('CLOSED', 'onclose', evt);
-    } else if (!this.unload) {
-      this._reconnect(evt, hasConnected);
-    }
-  }.bind(this);
-
-  /* ---------------------------------
-   * message
-   * -------------------------------*/
-  this.ws.onmessage = function (evt) {
-    this._trigger('onmessage', evt);
-  }.bind(this);
-
-  /* ---------------------------------
-   * error
-   * -------------------------------*/
-  this.ws.onerror = function (evt) {
-    this._trigger('onerror', evt);
-  }.bind(this);
-};
-
-
-/**
- * Wrapper around ws.send that adds data to queue if socket is not ready.
+ * @desc Wrapper around ws.send that adds queue functionality when socket is
+ *   not in a connected readyState.
+ *
+ * @example
+ * socket.send({ prop: 'val' });
  *
- * @public
  * @param {Object} data - data to send via web socket.
  */
 RcSocket.prototype.send = function (data) {
-  // If ready to send
+  // TODO: Seems like we should be checking if readyState is connected?
```

Comment:
Thats what it seems like to me too, `forcedClosed` too?

I think the key question is can we make this change with no impact to production.  Could we do a 'beta' release.